File: algorithms/algorithms-python/leetcode/Question_0071_Simplify_Path.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def simplifyPath(self, path):
        """
        :type path: str
        :rtype: str
        """
        items = path.split("/")
        res = []
        for i in items:
            if i == "." or not i:
                continue
            elif i == "..":
                if res:
                    res.pop()
            else:
                res.append(i)
        return "/" + "/".join(res)

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0071_Simplify_Path.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Question_0071_Simplify_Path {
    public String simplifyPath(String path) {
        if (path == null) {
            return path;
        }
        String[] items = path.split("/");
        Stack<String> stack = new Stack<>();
        for (String item : items) {
            if (item.isEmpty() || item.equals(".")) {
                continue;
            } else if (item.equals("..")) {
                if (stack.size() != 0) {
                    stack.pop();
                }
            } else {
                stack.push(item);
            }
        }
        if (stack.size() == 0) {
            return "/";
        }
        List<String> list = new ArrayList<>(stack);
        // 注意添加"/"
        return "/" + String.join("/", list);
    }
}
